File: leetcode-greedy-algorithms-solutions-optimized/src/main/java/com/greedyalgorithms/project/alternative/ActivitySelectionBruteForce.java
package com.greedyalgorithms.project.alternative;

import com.greedyalgorithms.project.utils.Activity;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Brute-force solution for the Activity Selection Problem.
 * This class explores all possible subsets of activities to find the maximum
 * number of non-overlapping activities. It demonstrates the inefficiency
 * of non-greedy approaches for this problem, contrasting with the
 * `GreedyAlgorithms.activitySelection` method.
 *
 * Time Complexity: O(2^N * N), where N is the number of activities.
 *                  Generating all subsets is O(2^N), and checking each subset
 *                  for non-overlap takes O(N) (after sorting each subset).
 * Space Complexity: O(N) for recursion stack and storing subsets.
 */
public class ActivitySelectionBruteForce {

    /**
     * Finds the maximum number of non-overlapping activities using a brute-force approach.
     * This method generates all possible subsets of activities and checks each subset
     * for non-overlapping properties, keeping track of the largest valid subset.
     *
     * @param activities A list of Activity objects.
     * @return A list of selected activities. Note: for brute force, the specific activities
     *         might differ from the greedy one if multiple optimal solutions exist,
     *         but the count will be the same.
     * @throws IllegalArgumentException if the input list is null.
     */
    public List<Activity> solve(List<Activity> activities) {
        if (activities == null) {
            throw new IllegalArgumentException("Activities list cannot be null.");
        }
        if (activities.isEmpty()) {
            return Collections.emptyList();
        }

        List<Activity> bestSolution = Collections.emptyList();
        // Generate all subsets and find the best one
        // The recursive helper `findSubsets` will do this.
        bestSolution = findSubsets(activities, 0, new ArrayList<>(), bestSolution);

        return bestSolution;
    }

    /**
     * Recursive helper to generate all subsets and find the one with maximum non-overlapping activities.
     * This method explores two branches for each activity: either include it in the current subset
     * or exclude it.
     *
     * @param allActivities The complete list of activities to choose from.
     * @param currentIndex The current index in `allActivities` being considered.
     * @param currentSubset The list of activities chosen for the current path.
     * @param bestGlobalSolution The best valid (non-overlapping) subset found so far.
     * @return The updated best valid subset.
     */
    private List<Activity> findSubsets(
            List<Activity> allActivities,
            int currentIndex,
            List<Activity> currentSubset,
            List<Activity> bestGlobalSolution) {

        // Base case: If all activities have been considered
        if (currentIndex == allActivities.size()) {
            // Check if the currentSubset is valid (non-overlapping)
            if (isNonOverlapping(currentSubset)) {
                // If it's valid and better than the best global solution, update.
                if (currentSubset.size() > bestGlobalSolution.size()) {
                    return new ArrayList<>(currentSubset); // Return a copy
                }
            }
            return bestGlobalSolution; // Return the current best
        }

        // Recursive Step 1: Exclude the current activity
        bestGlobalSolution = findSubsets(allActivities, currentIndex + 1, currentSubset, bestGlobalSolution);

        // Recursive Step 2: Include the current activity
        Activity activity = allActivities.get(currentIndex);
        currentSubset.add(activity); // Add it to the current path
        bestGlobalSolution = findSubsets(allActivities, currentIndex + 1, currentSubset, bestGlobalSolution);
        currentSubset.remove(currentSubset.size() - 1); // Backtrack: remove the activity for other paths

        return bestGlobalSolution;
    }

    /**
     * Checks if a given list of activities is non-overlapping.
     * This requires sorting the activities by start time for efficient checking.
     *
     * @param subset The list of activities to check.
     * @return true if all activities in the subset are non-overlapping, false otherwise.
     */
    private boolean isNonOverlapping(List<Activity> subset) {
        if (subset.size() <= 1) {
            return true;
        }

        // Sort the subset by start time to easily check for overlaps.
        // This is important because the subset might not be sorted.
        List<Activity> sortedSubset = new ArrayList<>(subset);
        sortedSubset.sort(Comparator.comparingInt(Activity::getStart));

        for (int i = 0; i < sortedSubset.size() - 1; i++) {
            if (sortedSubset.get(i).getFinish() > sortedSubset.get(i + 1).getStart()) {
                // An overlap found
                return false;
            }
        }
        return true;
    }
}

File: leetcode-greedy-algorithms-solutions-advanced/solutions_comparison/activitySelection_bruteForce.js
```javascript
/**
 * @fileoverview Brute-force (Recursive) solution for the Activity Selection Problem.
 *
 * This implementation explores all possible subsets of activities to find the
 * maximum number of non-overlapping activities. This is significantly less
 * efficient than the greedy approach but demonstrates a non-greedy way to solve
 * the problem for comparison.
 *
 * Problem Description:
 * Given N activities with start and finish times. Select the maximum number
 * of non-overlapping activities.
 *
 * Approach:
 * For each activity, we have two choices:
 * 1. Include the activity: If it doesn't overlap with previously selected activities,
 *    include it and recursively find the maximum activities from the remaining compatible ones.
 * 2. Exclude the activity: Recursively find the maximum activities from the remaining activities
 *    without including the current one.
 *
 * The `activitySelectionBruteForce` function initiates the process, potentially sorting
 * to make handling compatible activities slightly easier, but the core recursion still
 * explores all choices.
 */

/**
 * Recursive helper function for brute-force activity selection.
 *
 * @param {Array<Object>} activities - The array of all activities.
 * @param {number} index - The current activity index to consider.
 * @param {number} lastFinishTime - The finish time of the last selected activity.
 * @param {Array<Object>} currentSelection - The activities selected so far in the current path.
 * @param {Array<Object>} bestSelection - The best (maximum count) selection found so far.
 * @returns {Array<Object>} The overall best selection of activities.
 */
function findMaxActivitiesRecursive(activities, index, lastFinishTime, currentSelection, bestSelection) {
  // Base case: If all activities have been considered.
  if (index === activities.length) {
    // If the current selection is better than the best found so far, update best.
    if (currentSelection.length > bestSelection.length) {
      // Return a copy to avoid mutation by other recursive calls
      return [...currentSelection];
    }
    return bestSelection; // Return current best
  }

  // Option 1: Exclude the current activity
  let resultAfterExcluding = findMaxActivitiesRecursive(
    activities,
    index + 1,
    lastFinishTime,
    currentSelection,
    bestSelection
  );

  // Update bestSelection if resultAfterExcluding is better
  if (resultAfterExcluding.length > bestSelection.length) {
    bestSelection = resultAfterExcluding;
  }

  // Option 2: Include the current activity (if compatible)
  const currentActivity = activities[index];
  if (currentActivity.start >= lastFinishTime) {
    currentSelection.push(currentActivity);
    let resultAfterIncluding = findMaxActivitiesRecursive(
      activities,
      index + 1,
      currentActivity.finish, // Update last finish time
      currentSelection,
      bestSelection
    );
    currentSelection.pop(); // Backtrack: remove current activity for other paths

    // Update bestSelection if resultAfterIncluding is better
    if (resultAfterIncluding.length > bestSelection.length) {
      bestSelection = resultAfterIncluding;
    }
  }

  return bestSelection;
}

/**
 * Solves the Activity Selection problem using a brute-force recursive approach.
 *
 * @param {Array<Object>} activities - An array of activity objects.
 *   Each activity object must have 'start' and 'finish' properties (numbers).
 * @returns {Array<Object>} An array of selected activity objects.
 *
 * Time Complexity: O(2^N), where N is the number of activities.
 *                  In the worst case, for each activity, we explore two branches
 *                  (include or exclude). Sorting takes O(N log N) initially.
 *                  The exponential factor dominates.
 * Space Complexity: O(N) for recursion stack depth and storing intermediate selections.
 */
function activitySelectionBruteForce(activities) {
  if (!activities || activities.length === 0) {
    return [];
  }

  // Sorting by finish time initially can slightly prune the search space in some cases
  // but doesn't change the fundamental exponential complexity.
  // It's not strictly necessary for correctness of brute-force but good practice for any search.
  const sortedActivities = [...activities].sort((a, b) => a.finish - b.finish);

  const initialBestSelection = [];
  return findMaxActivitiesRecursive(sortedActivities, 0, -Infinity, [], initialBestSelection);
}

export default activitySelectionBruteForce;
```